pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java

```diff
@@ -663,6 +664,22 @@ public void release() {
         }
     }
 
+    @Override
+    public Optional<Long> getBrokerPublishTime() {
+        if (brokerEntryMetadata != null && brokerEntryMetadata.hasBrokerTimestamp()) {
+            return Optional.of(brokerEntryMetadata.getBrokerTimestamp());
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<Long> getIndex() {
+        if (brokerEntryMetadata != null && brokerEntryMetadata.hasIndex()) {
```

Comment:
For a batch message, we only have one index in the broker entry metadata, we will get the same index for the individual messages of a batch.

Reply:
Agree with @codelipenghui ,  for a batch message, we need to caculate the  individual index for each message from the index  of Entry and batch size

Reply:
Thank you all. Let me fix it.

Reply:
Fixed